File: MineAdmin-web/src/modules/base/api/menu.ts
/**
 * MineAdmin is committed to providing solutions for quickly building web applications
 * Please view the LICENSE file that was distributed with this source code,
 * For the full copyright and license information.
 * Thank you very much for using MineAdmin.
 *
 * @Author X.Mo<root@imoi.cn>
 * @Link   https://github.com/mineadmin
 */
import type { ResponseStruct } from '#/global'
import type { OperationRequest } from '@/generated/admin-api'

export interface MenuVo extends Partial<Omit<OperationRequest<'adminMenuCreate'>, 'btnPermission'>> {
  id?: number
  redirect?: string
  remark?: string
  btnPermission?: MenuVo[]
  [key: string]: any
}

export function page(): Promise<ResponseStruct<MenuVo[]>> {
  return useHttp().get('/admin/menu/list')
}

export function create(data: MenuVo): Promise<ResponseStruct<null>> {
  return useHttp().post('/admin/menu', toMenuPayload(data))
}

export function save(id: number, data: MenuVo): Promise<ResponseStruct<null>> {
  return useHttp().put(`/admin/menu/${id}`, toMenuPayload(data))
}

export function deleteByIds(ids: number[]): Promise<ResponseStruct<null>> {
  const data: OperationRequest<'adminMenuDelete'> = ids
  return useHttp().delete('/admin/menu', { data })
}

function toMenuPayload(data: MenuVo): Partial<OperationRequest<'adminMenuCreate'>> {
  return compactUndefined({
    parent_id: data.parent_id,
    name: data.name,
    path: data.path,
    component: data.component,
    redirect: data.redirect,
    status: data.status,
    sort: data.sort,
    meta: data.meta,
    btnPermission: data.btnPermission,
    remark: data.remark,
  })
}

function compactUndefined<T extends Record<string, unknown>>(value: T): Partial<T> {
  return Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined)) as Partial<T>
}

File: app/models/menu.go
package models

type Menu struct {
	ID        uint64  `gorm:"column:id;primaryKey" json:"id"`
	ParentID  uint64  `gorm:"column:parent_id" json:"parent_id"`
	Name      string  `gorm:"column:name" json:"name"`
	Meta      JSONMap `gorm:"column:meta;type:jsonb" json:"meta"`
	Path      string  `gorm:"column:path" json:"path"`
	Component string  `gorm:"column:component" json:"component"`
	Redirect  string  `gorm:"column:redirect" json:"redirect"`
	Status    int8    `gorm:"column:status" json:"status"`
	Sort      int16   `gorm:"column:sort" json:"sort"`
	AuditColumns
	Timestamps
	Remark string `gorm:"column:remark" json:"remark"`
}

func (Menu) TableName() string {
	return "menu"
}
